[#79] relative css injection improvements

Codex- opened the pull request:
This is a bit chunky, i've abstracted a few chunks out that can be isolated and tested independently, but have not added any tests yet.

Have a play, and see what you think @marco-prontera 

I imagine we'd want to move the abstracted methods into either `utils` or another file

Codex- commented:
From here, just need to decide if we want the exported functions in their own file or with the other utils, I also found an oddity where this uses esbuild with an explicit target instead of using the one from the vite config, but can sort that in another PR :)

marco-prontera commented on src/index.ts:
```diff
@@ -1,7 +1,107 @@
+    return cssSource instanceof Uint8Array ? new TextDecoder().decode(cssSource) : `${cssSource}`;
+}
+
+export function concatCss(bundle: OutputBundle, cssAssets: string[]): string {
```
We can name it `concatCssAndDeleteFromBundle` and delete inside that method instead. So `extractCssAndDeleteFromBundle` will become simple `extractCss`

Codex- replied:
Moved the deletion to the reducer

marco-prontera commented on src/index.ts:
```diff
@@ -45,107 +143,64 @@ export default function cssInjectedByJsPlugin({
                 });
+
+            if (relativeCSSInjection) {
+                const assetsWithCss = buildJsCssMap(bundle, jsAssetsFilterFunction);
+                await relativeCssInjection(bundle, assetsWithCss, buildCssCode);
```
After we call `buildJsCssMap` we are sure that every piece of CSS is injected somewhere? And no file.css is generated yet?

Codex- replied:
Any CSS that is marked as imported/included by an asset _should_ be injected to the relative importer, at this point any remaining CSS assets are probably kept, we could log them as warnings to make it visible when assets exist that have no parent?

marco-prontera replied:
Yes it's better to low a warning

Codex- replied:
Warning added, as well as an option to suppress the warning, as some consumers might have other parts of their bundling that handle these assets that are not handled by this plugin

marco-prontera reviewed (changes requested):
First of all, thank you!
I wrote some comments that can improve this thing a little bit, and after that, I think we are ready!

marco-prontera commented:
> From here, just need to decide if we want the exported functions in their own file or with the other utils, I also found an oddity where this uses esbuild with an explicit target instead of using the one from the vite config, but can sort that in another PR :)

I opened an issue to track the thing about the 'build.target'! 

I think we can write named functions instead of anonymous functions because it makes debugging more difficult. And moving all the functions from the top to the bottom of the file will make it a little more readable, do you agree?

Codex- commented:
> I opened an issue to track the thing about the 'build.target'!

Excellent, should be an easy fix :)
 
> I think we can write named functions instead of anonymous functions because it makes debugging more difficult. And moving all the functions from the top to the bottom of the file will make it a little more readable, do you agree?

Can do, did you want to keep these new functions in the file though? The problem is with keeping them there is that they're now exposed as part of the public API, ideally you want to import them from another file without re-exporting them and then only exporting the plugin / interfaces right?

Codex- commented:
> I think we can write named functions instead of anonymous functions because it makes debugging more difficult.

Which did you want to move? Any that are stored in a var actually return their name:

```ts
const someFunc = () => { throw new Error() };
```

```shell
> const someFunc = () => { throw new Error() };
undefined
> someFunc()
Uncaught Error
    at someFunc (REPL1:1:32)
> 
```
someFunc()

marco-prontera commented:
> > I opened an issue to track the thing about the 'build.target'!
> 
> Excellent, should be an easy fix :)
> 
> > And moving all the functions from the top to the bottom of the file will make it a little more readable, do you agree?
> 
> Can do, did you want to keep these new functions in the file though? The problem is with keeping them there is that they're now exposed as part of the public API, ideally you want to import them from another file without re-exporting them and then only exporting the plugin / interfaces right?

You are right, at this point, you can move also the functions that are now at the bottom also in the utils file.
For the anonymous functions question, I think it's better to give a name to functions used inside the `filter` and `map` calls (but it's not a problem)

Codex- commented:
> You are right, at this point, you can move also the functions that are now at the bottom also in the utils file. For the anonymous functions question, I think it's better to give a name to functions used inside the `filter` and `map` calls (but it's not a problem)

- Moved the helpers to `utils`
- Moved the tests to `utils.spec`
- Reorganised the existing tests a little.
- Updated some of the anonymous functions to be named.

marco-prontera commented:
> > You are right, at this point, you can move also the functions that are now at the bottom also in the utils file. For the anonymous functions question, I think it's better to give a name to functions used inside the `filter` and `map` calls (but it's not a problem)
> 
> * Moved the helpers to `utils`
> * Moved the tests to `utils.spec`
> * Reorganised the existing tests a little.
> * Updated some of the anonymous functions to be named.

Perfect!

Codex- commented:
We should probably add something like:

```
#### relativeCSSInjection (boolean)

Enabling `relativeCSSInjection` will identify which JS assets import which CSS, and inject the imported CSS only to the directly related JS assets.

Any CSS assets not associated with a JS asset will be logged at the end of `relativeCSSInjection`, however, this can be suppressed by also setting `suppressUnusedCssWarning` to `true`.
```

to the docs for this 🤔
